main/src/library/System.flix

```diff
@@ -30,8 +30,22 @@ namespace System {
         ///
         /// See also `Console.readLine` for reading from the console.
         ///
-        /// pub def readLines(): Iterator[String] & Impure = ???  // TODO
-
+        pub def readLines(): Iterator[String] & Impure =
+            import static get java.lang.System.in as getSystemIn;
+            import new java.io.InputStreamReader(##java.io.InputStream) as newInputStream;
+            import new java.io.BufferedReader(##java.io.Reader) as newBufferedReader;
+            import java.io.BufferedReader.readLine() as bfReadLine;
+
+            let br = newBufferedReader(newInputStream(getSystemIn()) as ##java.io.Reader); // up-cast to superclass
+            let nextLine = ref bfReadLine(br);
+            let done = () -> Object.isNull(deref nextLine);
+            let next = () -> {
+                let l = deref nextLine;
+                nextLine := bfReadLine(br);
```

Comment:
note that this line is potentially blocking. is this acceptable?

Reply:
It would be the same in Java right? So I think it's expected. As long as returning the iterator itself does not block.

Reply:
Yes thats true. The iterator itself cannot block... ish.
This iterator basically has a lookahead line since nexLine is loaded in advance. This is needed to check whether the stream has been consumed. So if the input stream doesn't have a full line to read (or is empty) and hasn't been closed, the readlines call itselfs blocks.

I think the issue here is that the underlying block doesn't really have a concept of `done`. It is for sure done when the stream is closed and we have read everything, but if we have read everything and the stream is open then the next readline might fail or not. we dont know.